[Testing] Adding reference tests to `OpInfo` class

krshrimali opened the pull request:
This PR will ideally add `ref` argument to `OpInfo` base class. The idea is to add reference checks for all the ops _eligible_. For more discussion, please check #58294

* [x] Migrate (but not removing yet) and modify helper functions from `UnaryUfuncOpInfo` class to `OpInfo` base class.
* [x] Test the reference checks for multiple ops. (also decide a list of different and eligible ops for this)
* [x] Handle possible edge cases (for example: `uint64` isn't implemented in PyTorch but is there in NumPy, and this needs to be handled -- more on this later) -- _Update_: We decided that these reference tests should only test for values and not types.
* [x] Create a sample PR for a single (of all different categories?) on adding reference functions to the eligible ops. -- _Update_: This is being done in this PR only.
* [x] ~Remove reference tests from `test_unary_ufuncs.py` and test to make sure that nothing breaks.~ (*Update*: We won't be touching Unary Ufunc reference tests in this PR)
* [x] Add comments, remove unnecessary prints/comments (added for debugging).

Note: To keep the PR description short, examples of edge cases encountered have been mentioned in the comments below.

cc: @mruberry @pmeier @kshitij12345

krshrimali committed:
Init-ing, op info ref

krshrimali committed:
Testing for roll

krshrimali committed:
Adding ref to OpInfo class

krshrimali committed:
import decorator

krshrimali committed:
Testing with test_funcs now

krshrimali committed:
Add sample_kwargs arg

krshrimali committed:
minor changes, import _NOTHING

krshrimali committed:
Testing further on power func

krshrimali committed:
Merge pull request #2 from krshrimali/origin_dev/opinfo_ref

Ready to be WIP in upstream

krshrimali commented:
Examples of few edge cases encountered so far:

* `torch.pow`, ref: `np.power`:

NumPy does not support integers as input and negative integers as powers (a `ValueError` is raised, see below). PyTorch, on the other hand, does not raise a warning/error but instead returns `tensor(0)`.

```
======================================================================
ERROR: test_reference_testing_pow_test_check_pow_cpu_int8 (__main__.TestOpInfoCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/Documents/Projects/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
  File "/home/krshrimali/Documents/Projects/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 266, in test_wrapper
    return test(*args, **kwargs)
  File "/home/krshrimali/Documents/Projects/Quansight/pytorch/torch/testing/_internal/common_utils.py", line 626, in wrapper
    fn(*args, **kwargs)
  File "/home/krshrimali/Documents/Projects/Quansight/pytorch/test/test_ops.py", line 132, in test_reference_testing
    self._test_reference_numerics(dtype, op, tensors)
  File "/home/krshrimali/Documents/Projects/Quansight/pytorch/test/test_ops.py", line 112, in _test_reference_numerics
    expected = op.ref(sample_numpy_input, *numpy_args, **numpy_kwargs)
ValueError: Integers to negative integer powers are not allowed.

----------------------------------------------------------------------
```

There are 2 possible solutions to this:

1. Don't do reference checking for `pow`. (OK for beginning, but there might be more ops who are divergent from NumPy's behavior, and this needs to be handled)
2. Add an option/decorator/argument to only run these tests for specific dtypes mentioned. (Covers more ops but will need some time spent on figuring out for which dtypes the implementation is divergent b/w NumPy and PyTorch)

(I can only think of these 2 right now, more suggestions and opinions are welcome).

* `torch.split`, ref: `np.split`:

The implementation of `split` is divergent compared to NumPy's `split`. See `#50012`. For such ops, I suggest to just not add reference checks.

For now, only the above 2 ops have shown divergent behavior in any way out of 3 tested (`split`, `pow` and `roll`). More ops to be tested.

kshitij12345 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -192,6 +192,8 @@ class OpInfo(object):
                  name,  # the string name of the function
                  *,
+                 ref=None,  # Just a reference
+                 sample_kwargs=lambda device, dtype, input: ({}, {}),
```
Note: This won't be needed as for this case as we are iterating over sample inputs which has kwargs field. 
In case of Unary, the inputs are generated by tests so we need a way mechanism to pass other arguments.

krshrimali replied:
Ah, interesting! Thanks, will remove this.

kshitij12345 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6505,6 +6514,7 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
 
 # Common operator groupings
 unary_ufuncs = [op for op in op_db if isinstance(op, UnaryUfuncInfo)]
+test_funcs = [op for op in op_db if isinstance(op, OpInfo)]
```
This won't filter out anything as all entries are derived from `OpInfo`. Just use `op_db` in `test_ops.py`

krshrimali replied:
True! Was checking what's the difference, I'll remove these unnecessary lines (also in the TODO). :) Thanks!

krshrimali replied:
I'm surprised I didn't remove this yet, apologies. I'll do remove this in the next commit.

mruberry replied:
Just be sure to keep conversations unresolved until the change appears to make it easier to follow the PR

kshitij12345 commented on test/test_ops.py:
```diff
@@ -45,6 +48,89 @@ def test_unsupported_dtypes(self, device, dtype, op):
+        else:
+            self.assertEqual(actual, expected, msg, exact_device=False, **kwargs)
+
+    def _test_reference_numerics(self, dtype, op, tensors, equal_nan=True):
```
Would be nice if the main logic of `_test_reference_numerics` and `assertEqualHelper` in this file and test_unary_ufunc is consolidated at a common place and then imported by these tests.

If we have to tweak anything in future, we'll have to do it in a single place.

krshrimali replied:
Oh, sorry. I missed adding this point in the TODO list, once migration is done and I'm done testing that it doesn't affect the Unary Ufuncs reference checks, then I'll bring them all to a single place. Thanks for pointing this out, @kshitij12345 !

mruberry replied:
More generally, I wonder if we should think about adding a helper function in common_utils.py that better supports testing against a NumPy reference. We have some old stuff but nothing modern.

kshitij12345 commented on test/test_ops.py:
```diff
@@ -21,6 +23,7 @@
 
 # Get names of all the operators which have entry in `method_tests` (legacy testing infra)
 method_tested_operators = set(map(lambda test_details: test_details[0], method_tests()))
+only_testing_ops = list(filter(lambda op: op.ref is not _NOTHING, test_funcs))
```
`ops_with_reference`?

krshrimali replied:
Sure, will make this change :) Thanks!

kshitij12345 reviewed (commented):
Thanks @krshrimali! Had a cursory look. Made a few comments and suggestions.

Apologies but I am a bit confused about the roadmap.

Will this PR try to add multiple operators or this PR plans to introduce the infrastructure and add reference for few operators and then gradually update the remaining operators in follow-up PRs? (Second one makes more sense to me.)

krshrimali commented:
> Will this PR try to add multiple operators or this PR plans to introduce the infrastructure and add reference for few operators and then gradually update the remaining operators in follow-up PRs? (Second one makes more sense to me.)

Thanks, it's a good question. I'll make the description clearer, but this is how I'm thinking the roadmap to be: (suggestions/feedback welcome)

- This PR will just add reference infrastructure to `OpInfo` class.
- But to have a useable infrastructure, I'll test this infra on a few ops (2-4) on my local system to make sure the infra covers comfortably enough cases. (I understand we'll know more once all the eligible ops use this infra). This will not be included in the PR though.
- Once this PR is merged, follow-up PRs will gradually update the OpInfo framework. (as you rightly mentioned)

Does this sound good? @kshitij12345

kshitij12345 commented:
> But to have a useable infrastructure, I'll test this infra on a few ops (2-4) on my local system to make sure the infra covers comfortably enough cases. (I understand we'll know more once all the eligible ops use this infra). This will not be included in the PR though.

Since you'll be testing them locally already it would make sense to add them to this PR. This would serve two purposes.
* It will actually test the new reference tests on CI.
* It will serve as an example to add these references so that other community members can also help you in this endeavor.

Otherwise sounds good to me. Thanks!

krshrimali committed:
Fix bugs, testing OK for roll and split ops

krshrimali committed:
Testing for add, added comments, minor changes

krshrimali committed:
More testing, take care of output dtypes

krshrimali committed:
Remove comments

krshrimali commented:
While testing on a other ops, here are a few more observations:

Consider `subtract` function of both NumPy and PyTorch: The `sample_inputs` func will usually return a combination of tensors and scalars for all the possible arguments of the op, one of the cases I found interesting is mentioned here:

```python
>>> torch.subtract(torch.tensor(1, dtype=torch.uint8), 3)
tensor(254, dtype=torch.uint8)
>>> np.subtract(np.array(1, dtype=np.uint8), 3)
-2
```

The assertion for both of them to be equal will obviously fail since both the values and dtypes are different. One solution would be to mention dtype in such ops using `np.subtract(1, 3, dtype=np.uint8)` but this doesn't seem to be an appropriate solution for these reasons:

1. Not all ops have the dtype argument in NumPy, an example: `np.amin`.
2. Explicitly passing output datatype for an op will essentially be disabling type assertions since anyways we pass the `dtype` to reference function.

Since I have noticed this as an issue with Scalars/Numbers, so one solution could be to convert each scalar to a tensor (with dtype as the input dtype, here `uint8`) and then depend on the library's type promotion logic to figure out the output dtype.

```python
>>> np.subtract(np.array(1, dtype=np.uint8), np.array(3, dtype=np.uint8))
254
```

But this solution won't work for the inputs: `torch.tensor(1, dtype=torch.bool)` and `1`. Since we plan to convert the scalar to tensor (for the solution proposed above), 1 will be converted to `torch.tensor(1, dtype=torch.bool)` (`True`). Considering `add` here (since `subtract` won't work on 2 boolean inputs on NumPy and PyTorch)

```python
>>> torch.add(torch.tensor(1, dtype=torch.bool), 1)
2
>>> np.add(np.array(1, dtype=torch.bool), np.array(1, dtype=torch.bool))
True
```

Here are a few ideas I have in my mind:

1. Only run the reference checks for non-scalar inputs. (have tested for multiple ops: `argmin, argmax, amin, amax, linalg.cholesky, dot, cidst -- scipy.spatial.distance.cdist`).
2. Disable tests for particular dtypes for each op. These tests can be particularly ignored for scalar inputs instead of ignoring for all inputs.

I'm oriented towards the 1st solution (to only run reference checks for non-scalar inputs), but there can be a better and logical solution to this, can't think of one right now. Any inputs/suggestions are welcome. @mruberry @pmeier @kshitij12345

krshrimali committed:
Ignore scalar inputs

krshrimali committed:
Add tests for more ops

krshrimali committed:
Merge branch 'master' into dev/opinfo_ref

krshrimali committed:
Merging with dev branch

krshrimali committed:
flake8 - indentation fix

mruberry commented:
Hey @krshrimali! Thanks for taking a close look at what references might look like, and what issues we may have when implementing references.

A few concepts for us to work with:

- a reference doesn't have to be a function in NumPy, it can be a function that wraps a NumPy (or SciPy) function, too, to implement the desired behavior. Sometimes NumPy ops don't work the way we'd expect.
- type promotion is a pain. In the future I expect we'll implement type promotion as specified by the Python Array API (https://data-apis.org/array-api/latest/API_specification/type_promotion.html?highlight=type%20promotion), but PyTorch's current type promotion is different. In the case of subtract above I think NumPy's type promotion might be divergent from the Python Array API's, too?
- instead of worrying about type promotion in this PR, maybe there's 1-2 simple references we can add that test the functionality this PR wants to implement?
- in the future, we might be able to deal with type promotion discrepancies by implementing a type promotion wrapper for NumPy operators. This might use torch.result_type, for example, and pre-cast all the inputs to the NumPy operation to the appropratie dtype.

Does that make sense? What are your thoughts?

mruberry commented on test/test_ops.py:
```diff
@@ -38,6 +43,89 @@ def test_unsupported_dtypes(self, device, dtype, op):
+            self.assertEqual(actual, expected, msg, exact_device=False, **kwargs)
+
+    def _test_reference_numerics(self, dtype, op, tensors, equal_nan=True):
+        def _helper_reference_numerics(expected, actual, msg, exact_dtype, equal_nan=True):
```
More correctly the test would have a name like, "test_reference" or "test_against_reference"

krshrimali replied:
Agreed. I like `_test_against_reference`.

mruberry commented on test/test_ops.py:
```diff
@@ -38,6 +43,89 @@ def test_unsupported_dtypes(self, device, dtype, op):
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype,
+                                       exact_dtype=exact_dtype, rtol=1e-3, atol=1e-2)
+
+            elif dtype is torch.bfloat16:
```
Since we know we're comparing with NumPy I think skipping bfloat16 by default is an option, if it makes things simpler

mruberry commented on test/test_ops.py:
```diff
@@ -38,6 +43,89 @@ def test_unsupported_dtypes(self, device, dtype, op):
+            else:
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype, equal_nan=equal_nan, exact_dtype=exact_dtype)
+
+        for sample_input in tensors:
```
Looks like the purpose of this block is to convert a sample input to a valid NumPy input, right? Instead of putting this logic here, what about extending the SampleInput class with a `.numpy()` method? It's OK if that method throws an error if an argument is a container for now.

krshrimali replied:
Having a `.numpy()` method will be neat, I agree. I was just curious if SciPy supports NumPy arrays as inputs, and looks like it does. (tested for a few ops in SciPy py passing `np.ndarray` input).

I'm sorry though, could you explain a little on what you meant by "if an argument is a container for now"? Do you mean we should only convert to a NumPy array if the input is a torch tensor or a scalar, and not if it's a list/tuple like containers? @mruberry

rgommers replied:
SciPy will also accept `torch.Tensor` instances, as long as they live on CPU and don't use autograd (SciPy will convert it to numpy.ndarray in its input parsing).

"container" is typically a list/tuple/set/etc.

krshrimali replied:
Thanks, @rgommers! Appreciate the help and clarification.

mruberry commented on test/test_ops.py:
```diff
@@ -38,6 +43,89 @@ def test_unsupported_dtypes(self, device, dtype, op):
+                for x, y in zip(expected, actual):
+                    _helper_reference_numerics(x, y, msg, exact_dtype, equal_nan)
+
+    @suppress_warnings
```
Add a comment for this test

krshrimali replied:
Yes, apologies. I'll add a comment here.

mruberry commented on test/test_ops.py:
```diff
@@ -38,6 +43,89 @@ def test_unsupported_dtypes(self, device, dtype, op):
+
+    @suppress_warnings
+    @ops(only_testing_ops)
+    def test_reference_testing(self, device, dtype, op):
```
If this PR updates the SampleInput class and adds a function common_utils for comparing with a NumPy reference, this function might look like:

```
@suppress_warnings
@ops(only_testing_ops, allowed_dtypes=(torch.float32, torch.long))
def test_reference_testing(self, device, dtype, op):
  for sample in op.sample_inputs(device, dtype):
    compare_with_numpy(op, op.ref, sample)
```

And then in common_utils.py the compare with NumPy function might look like:

```
def compare_with_numpy(op, ref, sample):
  sample_np = sample.numpy()
  actual = op(sample.splat())
  expected = op(sample.splat())
  self.assertEqual(actual, expected)
```

I cheated by also adding a `splat` method to sample inputs to help avoid having to parse out the input, arg, and kwargs all the time.

What would you think of a structure like this, @krshrimali, @kshitij12345?

krshrimali replied:
- I like the idea of having `compare_with_numpy` function and keeping the test neat and simple.
- Minor correction, I think you meant `expected = op.ref(sample_np.splat())` (just mentioning if someone reads this in future).

For `splat()` function, do you think it's better to have a static method `splat()` in the `SampleInput` class? The reason I say this is because `<SampleInputObject>.numpy()` should return a tuple of `(inputs, args, kwargs)` and not a `SampleInput` object. Hence it will not be possible to access `sample_np.splat()` (`smaple_np` being a tuple).

```python
actual = op(SampleInput.splat(sample))
expected = op.ref(SampleInput.splat(sample_np))
```

Having a splat function which takes an argument instead of `self`, is probably more useful? (please correct me if I'm wrong here)

```python ​
@staticmethd
def splat(sample):
   # inputs, args, kwargs (order)
   ​return tuple(sample[0], sample[1], sample[2],)
```

Instead of:

```python
# This will work if .numpy() returns a SampleInput object
def splat(self):
    return tuple(self.args, self.input, self.kwargs,)
```

_Note:_ the functions are just here for example, the actual implementation will be more accurate and different.

What are your thoughts on this? @mruberry @pmeier @kshitij12345

mruberry replied:
That's a great point. What about a `splat()` method that splats and a `numpy()` method that converts tensors into NumPy arrays but return a new SampleInput? Couldn't that be splatted?

krshrimali replied:
I did think of that as a solution but I am not sure if a splat function will work like what we expect it to be. Here is a workaround though:

```python
import torch

class SampleInput:
    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = c

    def fn(self, a):
        print(a)

    def splat(self, fn):
        return fn(self.a, self.b, **self.c,)

obj = SampleInput(torch.tensor(1), torch.tensor(2), {'alpha': 2})
print(obj.splat(torch.add))  # Returns tensor(5)
```

What do you think about this structure? @mruberry - Personally, I think it's reasonable as it does make things easier.

mruberry replied:
Seems like the idea of splat() isn't very interesting and we should just drop it from this PR

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -185,6 +185,8 @@ class OpInfo(object):
     def __init__(self,
                  name,  # the string name of the function
                  *,
+                 ref=None,  # Just a reference
```
Let's elaborate in this comment.

"An optional reference function that accepts NumPy arrays. If given, the op will be compared with its reference on each of its sample inputs."

krshrimali replied:
Ah, yes, I'm sorry. I understand I need to revise this comment, was just first focusing on the structure of changes. :) But thanks for taking a look and suggesting a better comment, Ruberry. :)

krshrimali replied:
Note to self: This is still in my TODOs, I'll work on adding/elaborating the comments once done with the design.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5900,6 +5906,7 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                        SkipInfo('TestUnaryUfuncs', 'test_out_arg_all_dtypes'),
                    )),
     OpInfo('roll',
+           ref=np.roll,
```
roll seems like a good candidate to start with, since it doesn't seem to have type promotion issues or other oddities that require writing a more involved reference

krshrimali replied:
Agreed!

mruberry reviewed (commented):
Looks like a good start, @krshrimali, I made some inline comments about structure that might be interesting. Looking forward to hearing your thoughts!

krshrimali commented:
> * a reference doesn't have to be a function in NumPy, it can be a function that wraps a NumPy (or SciPy) function, too, to implement the desired behavior. Sometimes NumPy ops don't work the way we'd expect.

Thanks for mentioning this. I initially started off with NumPy references, but later I realize there can be references from SciPy as well, or maybe in future from any other library. I'll also update the variable names to be more generic, and make sure that the implementation is enough flexible to have references from different libraries.

> * type promotion is a pain. In the future I expect we'll implement type promotion as specified by the Python Array API (https://data-apis.org/array-api/latest/API_specification/type_promotion.html?highlight=type%20promotion), but PyTorch's current type promotion is different. In the case of subtract above I think NumPy's type promotion might be divergent from the Python Array API's, too?

_[placeholder] - will update this comment after taking a look at the diagram properly_

> * instead of worrying about type promotion in this PR, maybe there's 1-2 simple references we can add that test the functionality this PR wants to implement?

This will definitely make things simpler, post offline discussion and as suggested by @pmeier, do you think converting the NumPy output using `torch.as_tensor(<val>, dtype=output_dtype_from_torch)` makes sense? I honestly think that it's a decent way to by pass the type promotions check, what are your views? @mruberry @kshitij12345 

> * in the future, we might be able to deal with type promotion discrepancies by implementing a type promotion wrapper for NumPy operators. This might use torch.result_type, for example, and pre-cast all the inputs to the NumPy operation to the appropratie dtype.

This will be a nice wrapper to have. :)

pmeier commented:
> converting the NumPy output using `torch.as_tensor(<val>, dtype=output_dtype_from_torch)` makes sense? I honestly think that it's a decent way to by pass the type promotions check

Note that I'm not advocating to not have type promotion checks in general, but we shouldn't do that here. For the references we should only check the values, since this thread as well as #58736 make clear that there is a lot of diverging behavior between the frameworks with respect to type promotion.

kshitij12345 commented:
> In the case of subtract, and the case I mentioned - it doesn't seem to be divergent from the Python Array API:
> 
> >>> np.subtract(np.array(1, dtype=np.uint8), np.array(3, dtype=np.int64)).dtype
> >>> dtype('int64')  # matches Python Array API's promotion: uint8, int64 ==> int64
> I think the divergence (not sure if it's the right word here) in this specific case for PyTorch is when a scalar is passed along with a tensor:
> 
> >>> torch.subtract(torch.tensor(1, dtype=torch.uint8), 3).dtype
> >>> torch.uint8
> 

Just a note:
The comparison is incorrect. For torch.subtract, you pass a tensor and a scalar while for numpy it is array and array.
For a like for like match, they agree in this case.
```python
>>> torch.subtract(torch.tensor(1, dtype=torch.uint8), torch.tensor(3, dtype=torch.int64)).dtype
torch.int64
>>> np.subtract(np.array(1, dtype=np.uint8), np.array(3, dtype=np.int64)).dtype
dtype('int64')
```
But I guess there are other discrepancies we already know of. Eg. https://github.com/pytorch/pytorch/issues/58736

> instead of worrying about type promotion in this PR, maybe there's 1-2 simple references we can add that test the functionality this PR wants to implement?



I agree that we should check for values and not the type promotion results. To that end however, I am of the opinion that the input to reference implementation should be casted to the result type of torch operation. 

This is mostly based on what @mruberry said
> in the future, we might be able to deal with type promotion discrepancies by implementing a type promotion wrapper for NumPy operators. This might use torch.result_type, for example, and pre-cast all the inputs to the NumPy operation to the appropratie dtype.

(And having been biten while working on UnaryUfunc)

Reason being operators in scipy.special particularly (I think many scipy ops will fit this) promote all integral inputs to the `double` while Pytorch would do it to `float` and do the computation. So this discrepancy of reference computing in `double` and torch computing in `float` may need us to relax the tolerance when it shouldn't be the case.
(With this we are skipping the type promotion of reference and essentially just checking the output)

```python
>>> scipy.special.xlogy(np.array([2], dtype=np.int32), np.array([2], dtype=np.int32)).dtype
dtype('float64') # int - int input
>>> scipy.special.xlogy(np.array([2], dtype=np.int32), np.array([2], dtype=np.int32))
array([1.38629436])

>>> scipy.special.xlogy(np.array([2], dtype=np.float32), np.array([2], dtype=np.float32)).dtype
dtype('float32') # cast input to float
>>> scipy.special.xlogy(np.array([2], dtype=np.float32), np.array([2], dtype=np.float32))
array([1.3862944], dtype=float32) # slight difference in output due to computation being in `float`
```
(Difference is acceptable in this case but in some cases it maybe more)

krshrimali commented:
> > In the case of subtract, and the case I mentioned - it doesn't seem to be divergent from the Python Array API:
> > > > > np.subtract(np.array(1, dtype=np.uint8), np.array(3, dtype=np.int64)).dtype
> > > > > dtype('int64')  # matches Python Array API's promotion: uint8, int64 ==> int64
> > > > > I think the divergence (not sure if it's the right word here) in this specific case for PyTorch is when a scalar is passed along with a tensor:
> > 
> > 
> > > > > torch.subtract(torch.tensor(1, dtype=torch.uint8), 3).dtype
> > > > > torch.uint8
> 
> Just a note:
> The comparison is incorrect. For torch.subtract, you pass a tensor and a scalar while for numpy it is array and array.
> For a like for like match, they agree in this case.
> 
> ```python
> >>> torch.subtract(torch.tensor(1, dtype=torch.uint8), torch.tensor(3, dtype=torch.int64)).dtype
> torch.int64
> >>> np.subtract(np.array(1, dtype=np.uint8), np.array(3, dtype=np.int64)).dtype
> dtype('int64')
> ```
> 
> But I guess there are other discrepancies we already know of. Eg. #58736
> 

Thanks, @kshitij12345 for the note but just to be clear - the comparison was for NumPy vs Python Array API and PyTorch vs Python Array API (instead of PyTorch vs NumPy). Hence:

When I say that PyTorch is divergent, I meant from the Python Array API (`uint8, int64` - should return `int64` but PyTorch returns `uint8` dtype output). I agree for both inputs as tensors, PyTorch does mathc the expected dtype, but the case seemed different for scalars, hence the statement.

> I agree that we should check for values and not the type promotion results. To that end however, I am of the opinion that the input to reference implementation should be casted to the result type of torch operation.
> 
> This is mostly based on what @mruberry said
> 
> > in the future, we might be able to deal with type promotion discrepancies by implementing a type promotion wrapper for NumPy operators. This might use torch.result_type, for example, and pre-cast all the inputs to the NumPy operation to the appropratie dtype.
> 
> (And having been biten while working on UnaryUfunc)
> 

Cool, was (curiously) testing this with scalars:

```python
# I'm considering the case of scalar as the second input here, either we can convert it to NumPy array or do explicit type casting like this:
>>> np.subtract(np.array(1, dtype=np.uint8), np.uint8(3)).dtype
>>> dtype('uint8')
```

Thanks for the inputs, @kshitij12345!

kshitij12345 commented:
> > Cool, was (curiously) testing this with scalars:
> 
> I'm considering the case of scalar as the second input here, either we can convert it to NumPy array or do explicit type casting like this:
> >>> np.subtract(np.array(1, dtype=np.uint8), np.uint8(3)).dtype
> >>> dtype('uint8')
> Thanks for the inputs, @kshitij12345!

Thanks for clarifying @krshrimali 
That makes sense to me and matches the behaviour as described [here](https://data-apis.org/array-api/latest/API_specification/type_promotion.html#mixing-arrays-with-python-scalars)

krshrimali committed:
Add numpy(), splat() functions, post-cast numpy output to torch output dtype

krshrimali committed:
Tested OK, workaround with splat func though

krshrimali committed:
Minor changes: comments and returning tuple

krshrimali committed:
We need a comment here, that's it

krshrimali committed:
Test for svd as well (scipy)

krshrimali committed:
numpy, splat methods, revise design, testing....

added numpy and splat methods, revise design

krshrimali committed:
Fix trailing whitespace

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into dev/opinfo_ref

krshrimali committed:
Remove spatial requirement for now

krshrimali committed:
Fixing flake8 error (overindentation)

krshrimali committed:
Fix flake8 errors

krshrimali committed:
Merge branch 'dev/opinfo_ref' of github.com:krshrimali/pytorch into dev/opinfo_ref

krshrimali commented on torch/testing/_internal/common_utils.py:
```diff
@@ -940,6 +940,35 @@ def get_comparison_dtype(a, b):
+    except TypeError:
+        # TODO: This needs a comment
+        return
+    t_inp, t_args, _ = sample_input.splat()
```
`torch_fn(sample_input.splat())` returns the following error. I am looking for suggestions on this.

```
======================================================================
ERROR: test_reference_testing_sub_test_subtract_cpu_int64 (__main__.TestOpInfoCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_device_type.py", line 266, in test_wrapper
    return test(*args, **kwargs)
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 627, in wrapper
    fn(*args, **kwargs)
  File "/home/krshrimali/git/krshrimali/pytorch/test/test_ops.py", line 149, in test_reference_testing
    compare_with_reference(op, op.ref, sample_input)
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_utils.py", line 942, in compare_with_reference
    torch_fn(*sample_input.splat())
  File "/home/krshrimali/git/krshrimali/pytorch/torch/testing/_internal/common_methods_invocations.py", line 331, in __call__
    return self.op(*args, **kwargs)
TypeError: sub() takes 2 positional arguments but 3 were given

----------------------------------------------------------------------
```

**Note:** Even if we are able to fix this, `n_kwargs` can't be passed and hence they are ignored to be `_`, an example is that `np.add` doesn't have `alpha` argument but `torch.add` has.

@mruberry @pmeier @kshitij12345 - Can you please take a loo at this?

mruberry replied:
Because this is trying to call sub like this:

`sub((a, b, {alpha=.5}))`

when it needs to be calling sub like this:

`sub(a, b, alpha=.5)`

For np.add it's just that np.add is not the correct reference for torch.add. It needs to be wrapped.

mruberry replied:
Don't get hung up on splatting, if it's a bad idea or it would take a significant amount of time to figure out that's really not what this PR is about

krshrimali commented on torch/testing/_internal/common_utils.py:
```diff
@@ -940,6 +940,35 @@ def get_comparison_dtype(a, b):
+    else:
+        msg = None
+
+    if isinstance(actual, torch.Tensor):
+        torch.testing.assert_close(actual, torch.as_tensor(expected, dtype=actual.dtype), msg=msg)
+    else:
+        for x, y in zip(expected, actual):
+            # Testing multi-outputs results
+            torch.testing.assert_close(actual, expected, msg=msg)
```
Note to self: This needs comment.

Note to reviewers: As per discussion, and also offline discussion with @pmeier - we want to limit this reference check (for now at least) to only test values and not types.

krshrimali replied:
Update: a `check_device=False` param has been added here to ensure no device checks are done for reference tests.

krshrimali reviewed (commented):
This is ready for another round of review. The things that have changed are:

1. Added `.numpy()` method to class `SampleInput`. Few notes about the function:
    * As per discussion, it raises `TypeError` if the sample isn't a torch tensor. 
    * Returns a tuple of `(input, args, kwargs)`.
2. Added `splat()` function - I'm a little skeptical of having this function. Since the purpose it was supposed to fulfill hasn't been implemented in this round of review, I've asked for suggestions though.
3. Update on strategy: we aren't testing types anymore, only values will be tested. The output from reference function will be casted to output type of torch function. (using `torch.as_tensor`)
4. Tested on more ops - successful.
5. Minor changes - change function names, added TODO for comments etc.

**Kind request:** Please review the design and implementation changes. Comments to the functions will be added, and lines which aren't used anymore will be removed in the next round.

Looking forward to your comments. Thanks!

**Update**:

1. Iterable tensors are also possible (e.g: `vstack`) and that case is now handled (https://github.com/pytorch/pytorch/pull/59369/files#r652101870): https://github.com/pytorch/pytorch/blob/18da8b71acf2570e07c3468a2bf0b7a8f063b845/torch/testing/_internal/common_methods_invocations.py#L141-L144
2. Reference checks for spectral ops and unary ufuncs are skipped in this PR and we rely on their individual reference check implementations. Please see: https://github.com/pytorch/pytorch/pull/59369/files#r652102407

**TODOs**:

Offline casual discussion with @kshitij12345 led to the following TODOs (list will be updated post review):

* [ ] Need to modify `sample_inputs` functions to not pass some specific kwargs for particular functions (like `np.add` doesn't have alpha arg but `torch.add` has), please see: https://github.com/pytorch/pytorch/pull/59369/files#r652105084.
* [ ] Figure out if `.splat()` is needed and is worth having (also a discussion point).

@mruberry @pmeier @kshitij12345 - PTAL, whenever you guys find time.

krshrimali committed:
Minor comments, and TODO

krshrimali committed:
Support iterable tensors, only ops with refs, still some failures

krshrimali committed:
Resolve merge conflict

krshrimali committed:
Remove unused imports

krshrimali committed:
Let UnaryUfuncs and SpectralOps do their thing, remove pow and cholesky

krshrimali committed:
Remove testing reference from test/test_ops

krshrimali committed:
Don't test for svd (edge case)

krshrimali commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -146,6 +147,33 @@ def formatter(arg):
+
+        if isinstance(sample_torch_input, torch.Tensor):
+            sample_numpy_input = sample_torch_input.cpu().numpy()
+        elif is_iterable_of_tensors(sample_torch_input):
+            sample_numpy_input = [sample_input.cpu().numpy() for sample_input in sample_torch_input]
+        else:
+            raise TypeError(f"Only torch.Tensor types support but got {type(sample_torch_input)}")
```
For functions like `vstack`, the input is list of tensors and thus this case has been handled here. I don't think there will be a case when the input will be a scalar but an else loop won't hurt here?

mruberry replied:
Just convert tensors to NumPy and preserve scalars/bools/strings etc. Handling containers in v0 is fine but be sure to call the conversion function recursively on containers. For now it could check for Sequences or Mappings.

krshrimali commented on test/test_ops.py:
```diff
@@ -38,6 +43,20 @@ def test_unsupported_dtypes(self, device, dtype, op):
+    @suppress_warnings
+    @ops(ref_test_ops, allowed_dtypes=(torch.float32, torch.long))
+    def test_reference_testing(self, device, dtype, op):
+        # TODO: We should ideally use these tests for UnaryUfuncInfo as well, see: https://github.com/pytorch/pytorch/pull/59369/
+        if isinstance(op, UnaryUfuncInfo) or isinstance(op, SpectralFuncInfo):
+            self.skipTest("Skipped! UnaryUfuncInfo and SpectralOpInfo have implemented reference checks separately.")
```
Since reference checks for Spectral Functions and Unary Functions is implemented separately and this PR isn't sufficient (in temrs of design) to address all the ops of both categories, I've added a skip to ensure that testing checks implemented in this PR aren't used for spectral/unary functions.

krshrimali commented on torch/testing/_internal/common_utils.py:
```diff
@@ -940,6 +940,35 @@ def get_comparison_dtype(a, b):
+    # Compares torch function with reference function for given sample input (object of SampleInput)
+    # Note: only values are compares, type comparison is not done here
+    try:
+        n_inp, n_args, _ = sample_input.numpy()
+    except TypeError:
+        # TODO: This needs a comment
+        return
+    t_inp, t_args, _ = sample_input.splat()
```
Note: post discussion with @kshitij12345 - the kwargs will be used from `sample_inputs` only but modifications will be done for specific ops which have divergence of torch vs ref functions (`torch.add` supports alpha param but `np.add` doesn't).

This PR will be updated with this change soon.

pmeier replied:
I think something like the `alpha` parameter warrants a separate `OpInfo` with a custom reference. For example:

```python
OpInfo(
    ...
    variant_test_name="with_alpha",
    # numpy has no builtin reference for the alpha kwarg, but it is easy enough to emulate
    ref=lambda input, other, *, alpha: np.add(input, np.multiply(alpha, other)),
    ...
)
```

mruberry replied:
Correct: we should write the correct reference using functions from NumPy (and SciPy)

krshrimali replied:
Great suggestions, for this PR - I'll probably avoid picking add and add a simple example. This OpInfo entry (for add with alpha) will be good for another PR, I guess.

krshrimali committed:
Dont' check device for reference checks

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4433,18 +4465,24 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    )),
     OpInfo('add',
+           ref=np.add,
+           variant_test_name='test_add',
```
I'm confused about why we need a `variant_test_name` here? Won't this name the test `test_test_add_...`? Also applies to the other occurrences below.

krshrimali replied:
Sorry for the confusion, this is just for testing to only test add function. This will be removed during the final touches in the PR once design review is done.

pmeier commented on torch/testing/_internal/common_utils.py:
```diff
@@ -940,6 +940,35 @@ def get_comparison_dtype(a, b):
+        msg = None
+
+    if isinstance(actual, torch.Tensor):
+        torch.testing.assert_close(actual, torch.as_tensor(expected, dtype=actual.dtype), msg=msg, check_device=False)
```
Since the tensor coming from numpy will always be on the CPU, I suggest we simply also move the original tensor to the CPU instead of disabling the check.
```suggestion
        torch.testing.assert_close(actual.cpu(), torch.as_tensor(expected, dtype=actual.dtype), msg=msg)
```

pmeier commented on torch/testing/_internal/common_utils.py:
```diff
@@ -940,6 +940,35 @@ def get_comparison_dtype(a, b):
+    if isinstance(actual, torch.Tensor):
+        torch.testing.assert_close(actual, torch.as_tensor(expected, dtype=actual.dtype), msg=msg, check_device=False)
+    else:
+        for x, y in zip(expected, actual):
+            # Testing multi-outputs results
+            torch.testing.assert_close(actual, torch.as_tensor(expected, dtype=actual.dtype), msg = msg, check_device=False)
```
`torch.testing.assert_close` has container support so there is no need to for this looping here. Plus, this is missing a check for the same amount of elements, since `zip` stops whenever one iterable is exhausted. So for example

```python
for x, y in zip([], [1, 2, 3]):
    assert False
```

passes just fine although it shouldn't.

krshrimali replied:
> torch.testing.assert_close has container support so there is no need to for this looping here

Cool, thanks for the info @pmeier! Will remove the looping.

> Plus, this is missing a check for the same amount of elements, since zip stops whenever one iterable is exhausted

Agreed, good point. I'll add an assertion here.

mruberry replied:
use assertEqual and make this a method on the testcase so it can inherit atol (and, in the future, rtol) from decorators like precisionOverride

mruberry replied:
(assertEqual also works on lists so I think you can just call it on the output)

pmeier replied:
In the future `self.assertEqual` will use `torch.testing.assert_close` internally, so we get all features from it.

pmeier reviewed (commented):
I've added some inline comments and questions.

krshrimali committed:
early exit if arg not a torch tensor (Suggestion from Philip)

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

krshrimali committed:
Move the original tensor to the CPU (suggestion from Philip)

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -146,6 +147,31 @@ def formatter(arg):
         return self._repr_helper(formatter)
 
+    def splat(self):
+        # TODO: Is this required? op(sample.splat()) doesn't work
```
What do we need to do to make splatting work in a reasonable way?

krshrimali replied:
I've added a comment here: https://github.com/pytorch/pytorch/pull/59369/files#r653451851

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -146,6 +147,33 @@ def formatter(arg):
+        else:
+            raise TypeError(f"Only torch.Tensor types support but got {type(sample_torch_input)}")
+
+        return (sample_numpy_input, numpy_args, sample_kwargs)
```
Kwargs can include tensors, too

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -940,6 +940,35 @@ def get_comparison_dtype(a, b):
+    actual = torch_fn(t_inp, *t_args, **t_kwargs)
+    expected = ref_fn(n_inp, *n_args, **n_kwargs)
+
+    # Crafts a custom error message for smaller, printable tensors
```
I wouldn't bother with a custom error message

krshrimali committed:
Addressing reviews, still testing

krshrimali committed:
Minor typo corrections, tested OK

krshrimali committed:
Merge branch 'master' into dev/opinfo_ref

krshrimali committed:
Remove splat method for now, all tests should pass

krshrimali committed:
Merge branch 'dev/opinfo_ref' of github.com:krshrimali/pytorch into dev/opinfo_ref

krshrimali committed:
Merge branch 'master' into dev/opinfo_ref

krshrimali committed:
Elaborate comment for ref

krshrimali committed:
sample_kwargs isn't used anymore

krshrimali committed:
Remove variant test names from each entry

krshrimali committed:
Add import for suppress_warnings

krshrimali committed:
Trailing whitespace removed

krshrimali committed:
Fix: handle list/tuple/scalars, ignore shapefuncinfo, modify sample inputs, remove argmax and argmin

krshrimali committed:
Merge branch 'dev/opinfo_ref' of github.com:krshrimali/pytorch into dev/opinfo_ref

krshrimali commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -147,6 +148,29 @@ def formatter(arg):
+            if isinstance(torch_inp, torch.Tensor):
+                numpy_inp = torch_inp.cpu().numpy()
+            elif is_iterable_of_tensors(torch_inp):
+                numpy_inp = np.array([t_inp.cpu().numpy() for t_inp in torch_inp])
```
Note: TODO

```
iter_class = type(torch_inp)
iter_class([to_numpy(t_inp) for t_inp in torch_inp])
```

krshrimali commented on test/test_ops.py:
```diff
@@ -165,6 +169,20 @@ def unsupported(dtype):
+    @ops(ref_test_ops, allowed_dtypes=(torch.float32, torch.long))
+    def test_reference_testing(self, device, dtype, op):
+        # TODO: We should ideally use these tests for UnaryUfuncInfo as well, see: https://github.com/pytorch/pytorch/pull/59369/
+        if isinstance(op, UnaryUfuncInfo) or isinstance(op, SpectralFuncInfo) or isinstance(op, ShapeFuncInfo):
```
TODO:

`if isinstance(op, (UnaryUfuncInfo, SpectralFuncInfo, ShapeFuncInfo))`

krshrimali committed:
final flake8 and mypy fix

krshrimali committed:
Add lambda wrapper around ref func

krshrimali committed:
Add type annotation for mypy

krshrimali committed:
Just ignore type check now

krshrimali committed:
Hoping this will fix mypy

krshrimali committed:
Bring back iterable tensors check, use exact_device - linalg checks won't work for now on empty tensors

krshrimali committed:
mypy will anyways fail, let's take a look later. flake8 should be OK now

krshrimali committed:
Minor syntax fix

krshrimali commented on test/test_ops.py:
```diff
@@ -165,6 +169,20 @@ def unsupported(dtype):
+    @suppress_warnings
+    @ops(ref_test_ops, allowed_dtypes=(torch.float32, torch.long))
+    def test_reference_testing(self, device, dtype, op):
+        # TODO: We should ideally use these tests for UnaryUfuncInfo as well, see: https://github.com/pytorch/pytorch/pull/59369/
```
Note to self: Need to remove this TODO as we aren't sure if it's really a necessity/high priority.

krshrimali commented on test/test_ops.py:
```diff
@@ -165,6 +169,20 @@ def unsupported(dtype):
+    def test_reference_testing(self, device, dtype, op):
+        # TODO: We should ideally use these tests for UnaryUfuncInfo as well, see: https://github.com/pytorch/pytorch/pull/59369/
+        if isinstance(op, (UnaryUfuncInfo, SpectralFuncInfo, ShapeFuncInfo)):
+            self.skipTest("Skipped! UnaryUfuncInfo and SpectralOpInfo have implemented reference checks separately.")
```
Note to self: improve this message, and also add `ShapeFuncInfo` if keyword if required.

krshrimali commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -147,6 +148,26 @@ def formatter(arg):
+            # Input types supported: tensor/iterable tensors/list/tuple/dict/scalars
+            # For iterable tensors/list/tuple/dict - to_numpy() is recursively called trusting the end value will be
+            # either a torch tensor or a scalar
+            if isinstance(torch_inp, torch.Tensor):
+                numpy_inp = torch_inp.cpu().numpy()
+            elif is_iterable_of_tensors(torch_inp) or isinstance(torch_inp, (list, tuple)):
+                iter_class = type(torch_inp)
+                numpy_inp = iter_class(to_numpy(t_inp) for t_inp in torch_inp)	
+            elif isinstance(torch_inp, dict):
+                numpy_inp = {k: to_numpy(t_inp) for k, t_inp in torch_inp.items()}
+            else:
+                numpy_inp = torch_inp
```
As per suggestion, for now sequences and mapppings have been handled. For scalars, we return them as it is.

mruberry replied:
What about this for the internal helper?

```
# Converts tensors to ndarrays by calling .detach().cpu().numpy() on them
# Numbers, strings, and bool are preserved as is
# Lists, tuples and dicts are handled by calling this function recursively
def to_numpy(x):
  def _np(t):
    return t.detach().cpu().numpy()

  if isinstance(x, torch.Tensor):
    return _np(x)
  elif isinstance(x, list):
    return list(map(to_numpy, x))
  elif isinstance(x, tuple):
    return tuple(map(to_numpy, x))
  elif isinstance(x, dict):
    return {k: to_numpy(v) for k, v in x.items()}
  elif isinstance(x, (numbers.Number, bool, str)):
    return x
  
  raise ValueError("Unknown type {0}!".format(type(x)))
```

krshrimali replied:
Nice, thanks @mruberry - added in the recent commit.

krshrimali commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1564,20 +1588,20 @@ def sample_inputs_take_along_dim(op_info, device, dtype, requires_grad, **kwargs
-        ((), (0,)),
-        ((), ()),
-        ((), (0, True,)),
+    test_cases: Tuple[tuple, tuple, dict] = (  # type: ignore[assignment]
+        ((S, S, S), (), {}),
+        ((S, S, S), (1,), {}),
+        ((S, S, S), ((1, 2,),), {}),
+        ((S, S, S), (1,), {'keepdim': True}),
+        ((), (0,), {}),
+        ((), (), {}),
+        ((), (0,), {'keepdim': True}),
```
**Why this change?**

Example function: `np.amax` (https://numpy.org/doc/stable/reference/generated/numpy.amax.html)

- NumPy takes `out` as a keyword argument while PyTorch takes `out` as kw-only argument.
- NumPy function's signature has `out` as the 3rd argument (/or an argument which doesn't match the position of `keepdim` from `sample_inputs`).

For example, here, if the earlier case is considered, and we call the reference function: `ref_fn(inp, *args, **kwargs)` (where kwargs is `{}`), then boolean is going to be passed to `out=` param in NumPy's amax function call. Which is incorrect and will throw an error.

(continuing from the points above)

- OpInfo already has ability to handle `kwargs` from sample inputs functions.
- Since `keepdim` is a keyword argument in PyTorch, so it does make sense to pass as a kwarg.
- This will resolve the above error of function call assuming `keepdim` values passed as `out=` values.

(Tested OK)

**Notes**:

- For future PRs on adding reference functions, this change will be desired if there is `keepdim` like argument.

krshrimali commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4659,9 +4685,11 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                SkipInfo('TestCommon', 'test_variant_consistency_eager'),),
            sample_inputs_func=sample_inputs_addcmul_addcdiv),
     OpInfo('amax',
+           ref=lambda a, dim=None, keepdim=False, **kwargs: np.amax(a, axis=dim, keepdims=keepdim, **kwargs),
```
- NumPy takes `keepdims` as an argument while PyTorch takes `keepdim`.
- From the above comment (todo: update link), we pass `keepdim` as keyword argument but since NumPy doesn't have it, hence wrapping the function in a lambda and passing `keepdim` value to `keepdims` is recommended solution.

mruberry replied:
This is correct. References should have the same signatures as the PyTorch functions (when required).

Separately, in the future we should update our arg parsing to accept both keepdim and keepdims. I'll file some tracking issues.

krshrimali commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1128,6 +1128,28 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+                  f" got {len(actual)} from torch and {len(expected)} from NumPy."
+            assert len(actual) == len(expected), msg
+            for x, y in zip(actual, expected):
+                self.assertEqual(x, torch.as_tensor(y, dtype=x.dtype), exact_device=False)
```
**Why looping?** (`for x, y in zip(actual, expected)`)

- `actual.dtype` will not work if actual is an iterable of tensors/iterator. (For example: `torch.linalg.qr` returns tuple of tensors)
- `actual.device` will also not work (same reason as above)
- Looping is required since we are doing `torch.as_tensor(y, dtype=x.dtype)` for each tensor in the iterator of tensors.

Note: `actual.cpu()` will also not work for iterable of tensors, hence `exact_device=False`. For consistency, other conditions also use `exact_device=False`.

krshrimali reviewed (commented):
**Changes made:**

- Test skips for `ShapeFuncInfo` as well now. (they have ref tests implemented)
- Lambda wrapper suggested for reference functions having `keepdims` (NumPy) as an arg. (https://github.com/pytorch/pytorch/pull/59369#discussion_r654832385)
- Modify sample inputs functions to take `keepdims` as a kwarg instead of an arg. (https://github.com/pytorch/pytorch/pull/59369#discussion_r654832635)
- Added looping for output as iterbale of tensors, and used `exact_device=False` in `assertEqual` instead of `actual.cpu()`. (https://github.com/pytorch/pytorch/pull/59369#discussion_r654833158) 

**TODOs:**

* [x] Only keep 2 references (one with lambda wrapper and one without lambda wrapper) in this PR.
* [x] Handle `Note to self` review comments above. (minor changes)
* [x] Add more comments.
* [x] Final flake8, mypy and linter fixes.

cc: @mruberry @pmeier @kshitij12345 - Please take a look whenever you find time, thanks!

krshrimali committed:
Minor updates, remove TODOs and scipy import

mruberry commented on test/test_ops.py:
```diff
@@ -24,6 +24,10 @@
 
+# Get names of all the operators which have ref in their entry in OpInfo (testing infra)
+# TODO (@krshrimali): Add None to _NOTHING?
+ref_test_ops = list(filter(lambda op: op.ref is not None and op.ref is not _NOTHING, op_db))
```
Also filter out the UnaryUfuncInfos, since they're tested against their references in test_unary_ufuncs.py

krshrimali replied:
Done, this was done separately but now it's all done in the same place.

mruberry commented on test/test_ops.py:
```diff
@@ -165,6 +169,20 @@ def unsupported(dtype):
 
         self.assertEqual(supported_backward_dtypes, claimed_backward_supported, msg=msg)
 
+    # Tests that the function and its (array-accepting) reference produce the same
```
"array" -> "ndarray" to clarify that it's a NumPy array that the references accept

krshrimali replied:
Fixed in the recent commit, thanks!

mruberry commented on test/test_ops.py:
```diff
@@ -165,6 +169,20 @@ def unsupported(dtype):
+    @ops(ref_test_ops, allowed_dtypes=(torch.float32, torch.long))
+    def test_reference_testing(self, device, dtype, op):
+        # TODO: We should ideally use these tests for UnaryUfuncInfo as well, see: https://github.com/pytorch/pytorch/pull/59369/
+        if isinstance(op, (UnaryUfuncInfo, SpectralFuncInfo, ShapeFuncInfo)):
```
I would just filter these out in the initial set (see note above).

Where are the SpectralFuncInfo and ShapeFuncInfo tests vs their references?

krshrimali replied:
For `ShapeFuncInfo` ops (there are only 2 funcs: `repeat` and `tile`): https://github.com/pytorch/pytorch/blob/5ec4ad7f54cc35ee40b5991bd65c974056048f84/test/test_shape_ops.py#L681

For `SpectralFuncInfo` ops: (all `fft` namespace ops):
https://github.com/pytorch/pytorch/blob/5ec4ad7f54cc35ee40b5991bd65c974056048f84/test/test_spectral_ops.py#L102

mruberry replied:
Ah, OK. Only the 1D spectral ops are compared with their reference today, but it's fine to exclude them for now with a note. 

I think TestShapeFuncs can be deleted in favor of this new test, right? Doesn't it do the same thing?

krshrimali replied:
Agreed, `TestShapeFuncs` can be deleted - just that they test for multiple dtypes, I think just testing for `torch.float32, torch.long` is OK. Went through the documentation, didn't see anything specific to convince to test for complex types as well.

The dtypes the reference test for shape functions allow (currently) are here: https://github.com/pytorch/pytorch/blob/5ec4ad7f54cc35ee40b5991bd65c974056048f84/test/test_shape_ops.py#L679-L681

Just curious though, do we plan to add more ops in `ShapeFuncInfo` (which should not be reference tested) because there is an additional check in the test in `TestShapeFuncs`: https://github.com/pytorch/pytorch/blob/5ec4ad7f54cc35ee40b5991bd65c974056048f84/test/test_shape_ops.py#L680.

In the recent commit: I've added a note (for Spectral Funcs and Unary Ufuncs) and deleted `TestShapeFuncs` & added these funcs to new infra.

mruberry replied:
We may add more in the future but let's not worry about it. As long as tile and repeat have their references tested by the new test we don't need to bother with the old test and it can be deleted.

krshrimali replied:
Great, thanks for the clarification @mruberry!

mruberry commented on test/test_ops.py:
```diff
@@ -165,6 +169,20 @@ def unsupported(dtype):
+        if isinstance(op, (UnaryUfuncInfo, SpectralFuncInfo, ShapeFuncInfo)):
+            self.skipTest("Skipped! UnaryUfuncInfo and SpectralOpInfo have implemented reference checks separately.")
+
+        sample_inputs = op.sample_inputs(device, dtype)
```
Body of this test looks good

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -147,6 +148,26 @@ def formatter(arg):
+            # either a torch tensor or a scalar
+            if isinstance(torch_inp, torch.Tensor):
+                numpy_inp = torch_inp.cpu().numpy()
+            elif is_iterable_of_tensors(torch_inp) or isinstance(torch_inp, (list, tuple)):
```
I wouldn't use this function.

krshrimali replied:
Fixed.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -186,6 +207,8 @@ class OpInfo(object):
     def __init__(self,
                  name,  # the string name of the function
                  *,
+                 ref=None,  # An optional reference function that accepts NumPy arrays.
```
"NumPy" -> "ndarrays (AKA "NumPy arrays")"

krshrimali replied:
Fixed in the recent commit, thanks!

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1564,20 +1588,20 @@ def sample_inputs_take_along_dim(op_info, device, dtype, requires_grad, **kwargs
-        ((), (0,)),
-        ((), ()),
-        ((), (0, True,)),
+    test_cases: Tuple[tuple, tuple, dict] = (  # type: ignore[assignment]
```
Document what each of these tuples is. Shape, positional args, kwargs?

krshrimali replied:
Done, fixed in the recent commit. Thanks!

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1128,6 +1128,28 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+
+        actual = torch_fn(t_inp, *t_args, **t_kwargs)
+        expected = ref_fn(n_inp, *n_args, **n_kwargs)
+
```
I think this can just do `self.assertEqual(actual, expected, exact_device=False)`

krshrimali replied:
Makes sense, passes all the tests on local server. Thanks!

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1128,6 +1128,28 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+                self.assertEqual(x, torch.as_tensor(y, dtype=x.dtype), exact_device=False)
+        else:
+            # TODO: Does PyTorch eer return other than tensor/iterable of tensors?
+            raise TypeError(f"Got type: {type(actual)} from PyTorch but only tensors/iterable of tensors are currently supported")
```
I think this can be removed, too

krshrimali replied:
Done! Thanks.

mruberry commented:
Made some simplifying comments, @krshrimali. Once they're incorporated I think this PR should be close to being ready.

krshrimali committed:
Addressing review

krshrimali committed:
Merge branch 'dev/opinfo_ref' of github.com:krshrimali/pytorch into dev/opinfo_ref

krshrimali committed:
Address reviews

krshrimali committed:
Only test for amax, amin and roll

krshrimali commented:
> Made some simplifying comments, @krshrimali. Once they're incorporated I think this PR should be close to being ready; be sure to check that the lint issue is resolved, too. Other failures look unrelated.

Thanks @mruberry for the review. I'e addressed the comments, and revised this PR testing for `amax, amin, roll` (`amax, amin` cover lambda wrapper example while roll is a simple op with `np.roll` reference). Waiting for the CI to pass.

krshrimali committed:
Remove TestShapeFuncs (now done in new ref test framewor)

krshrimali committed:
minor flake8 fix

krshrimali committed:
Unused imports removed

mruberry commented:
Remember to flake8 locally before submitting with a command like `python -m flake8 <file>`.

This has a conflict that needs to be resolved

krshrimali committed:
Merge branch 'master' into dev/opinfo_ref

krshrimali commented:
> Remember to flake8 locally before submitting with a command like `python -m flake8 <file>`.

Sure, thanks for the suggestion! Will take care of this.

> This has a conflict that needs to be resolved

Thanks, I've resolved this @mruberry.

mruberry commented:
Mac build should be fixed now; would you rebase this to retrigger that and the other tests? Sorry about that

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into dev/opinfo_ref

krshrimali commented:
> Mac build should be fixed now; would you rebase this to retrigger that and the other tests? Sorry about that

Done, thanks @mruberry! :)

krshrimali commented:
The test failure looks unrelated to this PR. Rest of the tests seem to pass, please let me know if there are any changes suggested whenever you find time. @mruberry Thanks! :)

mruberry reviewed (approved):
Nice work, @krshrimali!

krshrimali commented:
Thanks @mruberry, @kshitij12345, @pmeier and @rgommers for helping in reviewing the PR. 🎉
